pkg/util/grpcutil/grpc_util_test.go

```diff
@@ -153,13 +53,19 @@ func TestRequestDidNotStart_Errors(t *testing.T) {
 		err    error
 		expect bool
 	}{
-		"breaker":          {errors.Wrapf(circuit.ErrBreakerOpen, "unable to dial n%d", 42), true},
-		"waiting for init": {errors.Wrapf(server.NewWaitingForInitError("foo"), "failed"), true},
-		"plain":            {errors.New("foo"), false},
+		"failed heartbeat":    {&netutil.InitialHeartbeatFailedError{}, true},
+		"waiting for init":    {server.NewWaitingForInitError("foo"), true},
+		"unauthenticated":     {status.Error(codes.Unauthenticated, "unauthenticated"), true},
+		"permission denied":   {status.Error(codes.PermissionDenied, "permission denied"), true},
+		"failed precondition": {status.Error(codes.FailedPrecondition, "failed precondition"), true},
+		"circuit breaker":     {circuit.ErrBreakerOpen, true},
```

Comment:
this one could be dangerous, right? If you send a write through DistSender, and it actually does stuff on a few ranges, but then hits a _replica_ circuit breaker error somewhere else, DistSender might think "request did not start" but it very much started indeed.
I see that we're only using `RequestDidNotStart` when sending to a single range, so it's probably... truer? But we could also be getting this error from another replica during intent resolution (see your PR from yesterday). I don't see how it can cause problems, exactly, but it seems a little unclean.

Reply:
Note that `circuit` here points to the old RPC circuit breaker package (`cockroachdb/circuitbreaker`), not your new circuit breakers.

But considering you recently refactored RPC to use the same circuit breaker infrastructure, I'd say yeah, it's probably worth revisiting this stuff. I think this probably doesn't work anymore at all, because we don't return that circuit breaker error.

Furthermore, the RPC circuit breaker error will be caught by the dial and promoted to an `InitialHeartbeatFailedError`, which is really the error we're interested in. I think we can probably just remove the circuit breaker case here, and it'll be good.

In any case, there are no logic changes in this PR, so it's sort of orthogonal. Do you mind following up on this and making sure there aren't any bad interactions?

Reply:
Ok, filed an issue.